euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Join.java

```diff
@@ -307,99 +305,179 @@ public static OfBuilder named(String name) {
       boolean leftEmpty = !lefts.iterator().hasNext();
       boolean rightEmpty = !rights.iterator().hasNext();
-      if (leftEmpty ^ rightEmpty) {
-        // if just a one collection is empty
+      // if just a one collection is empty
+      if (leftEmpty != rightEmpty) {
         if (leftEmpty) {
           for (RIGHT elem : rights) {
-            functor.apply(null, elem, getContext());
+            functor.apply(null, elem, context);
           }
         } else {
           for (LEFT elem : lefts) {
-            functor.apply(elem, null, getContext());
+            functor.apply(elem, null, context);
           }
         }
       }
     }
+  }
+
+  /**
+   * An implementation of the join state which will accumulate elements
+   * until it is flushed at which point it then emits all elements.<p>
+   *
+   * (This implementation is known to work correctly with merging
+   * windowing, early triggering, as well as with timed multi-window
+   * windowing (e.g. time sliding.))
+   */
+  private class StableJoinState extends AbstractJoinState
+      implements StateSupport.MergeFrom<StableJoinState> {
+
+    StableJoinState(StorageProvider storageProvider) {
+      super(storageProvider);
+    }
+
+    @Override
+    public void add(Either<LEFT, RIGHT> elem) {
+      if (elem.isLeft()) {
+        leftElements.add(elem.left());
+      } else {
+        rightElements.add(elem.right());
+      }
+    }
+
+    @Override
+    public void flush(Context<OUT> context) {
+      Iterable<LEFT> lefts = leftElements.get();
+      Iterable<RIGHT> rights = rightElements.get();
+      for (LEFT l : lefts) {
+        for (RIGHT r : rights) {
+          functor.apply(l, r, context);
+        }
+      }
+      if (outer) {
+        flushUnjoinedElems(context, lefts, rights);
+      }
+    }
+
+    @Override
+    public void mergeFrom(StableJoinState other) {
+      this.leftElements.addAll(other.leftElements.get());
+      this.rightElements.addAll(other.rightElements.get());
+    }
+  }
+
+  /**
+   * An implementation of the join state which produces results, i.e. emits
+   * output, as soon as possible. It has at least the following short commings
+   * and should be used with care:
+   *
+   * <ul>
+   *   <li>This implementation will break the join operator if used with a
+   *        merging windowing strategy, since items will be emitted under the
+   *        hood of a non-final window.</li>
+   *   <li>This implementation cannot be used together with early triggering
+   *        on any windowing strategy as it will emit each identified pair
+   *        only once during the whole course of the state's life cycle.</li>
+   *   <li>This implementation will also break time-sliding windowing, as
+   *        it will raise the watermark too quickly in downstream operators,
+   *        thus, marking earlier - but actually still not too late time-sliding
+   *        windows as late comers.</li>
```

Comment:
yes, you exactly hit the point regarding the watermark issue! i will have to investigate around this as i fear it'll break attached windowing. anyway i would do such a change separately and will set up a ticket for that.